HHH-12973 - Inconsistent identity generation

vladmihalcea opened the pull request:
https://hibernate.atlassian.net/browse/HHH-12973

vladmihalcea commented on hibernate-core/src/main/java/org/hibernate/internal/CoreMessageLogger.java:
```diff
@@ -1831,4 +1831,8 @@ void attemptToAssociateProxyWithTwoOpenSessions(
+
+	@LogMessage(level = WARN)
+	@Message(value = "The increment size of the [%s] sequence is set to [%d] in the entity mapping while the associated database sequence increment size is [%d]. The database sequence increment size will take precedence to avoid identifier  allocation conflicts.", id = 494)
+	void sequenceIncrementSizeMismatch(String sequenceName, int incrementSize, int databaseIncrementSize);
```
Good catch.

vladmihalcea commented on hibernate-core/src/main/java/org/hibernate/id/enhanced/SequenceStyleGenerator.java:
```diff
@@ -521,4 +572,39 @@ public String determineBulkInsertionIdentifierGenerationSelectFragment(Dialect d
+			try (PreparedStatement statement = connection.prepareStatement(
+					"SELECT increment " +
+							"FROM information_schema.sequences " +
+							"WHERE sequence_name = ?" )) {
```
I'll have to check to see how to accommodate `sequence_catalog` and `sequence_schema` so that it works on all major DBs.

sebersole replied:
Some databases do in fact allow having same named sequences in different namespaces

vladmihalcea replied:
I tested it on Oracle, SQL Server, PostgreSQL and H2 and it worked on all of them as the Information_Schema should, theoretically, be defined by the SQL Standard. However, it could be that some DBs will not honor the same structure, in which case, the request will not break since we don't escalate the exception up the stack. It would just be that Hibernate cannot determine the increment size. Nevertheless, I could push this query into the base Dialect, so we could override it for those DBs which don't follow the SQL Standard guidelines for Information Schema.

vladmihalcea commented:
@gsmet If the allocation size is 50 in Hibernate but the DB sequence allocation size is 1, then we indeed get that weird behavior described by this issue. Hibernate will assume that it can generate the identifiers according to pooled optimizer. But for it to work properly, we need both Hibernate and the DB to agree on the same allocation size.

sebersole commented:
> @gsmet If the allocation size is 50 in Hibernate but the DB sequence allocation size is 1, then we indeed get that weird behavior described by this issue. Hibernate will assume that it can generate the identifiers according to pooled optimizer. But for it to work properly, we need both Hibernate and the DB to agree on the same allocation size.

Correct - but this would be user/config error

vladmihalcea commented:
@sebersole @gsmet I extracted the sequence info retrieval at the Dialect level, and I added support and tested it for:

- Oracle
- SQL Server
- PostgreSQL
- DB2
- HANA
- Informix
- H2
- HSQLDB

MariaDB 10.3 does not support the INFORMATION_SCHEMA.SEQUENCES yet, it will be there in 10.4.

Let me know what you think.

sebersole commented:
I'm not sure what you mean - sequence info retrieval is already delegated to the Dialect : `org.hibernate.dialect.Dialect#getSequenceInformationExtractor`. 

Having "the DB increment to take precedence over the annotation-based one" is a slippery slope.  I'll reach out to you on Hip Chat to discuss why...

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/enhanced/SequenceStyleGenerator.java:
```diff
@@ -341,9 +389,10 @@ protected Identifier determineValueColumnName(Properties params, JdbcEnvironment
 	 * @return The initial value
 	 */
-	@SuppressWarnings({"WeakerAccess"})
+	@SuppressWarnings({ "WeakerAccess" })
```
Is your IDE set to auto-change lines?  Or did you manually make all of these kind of changes?

vladmihalcea replied:
I run a CTRL+ALT+L, so it applied these changes for all the code in this class.

typedt commented on hibernate-core/src/main/java/org/hibernate/id/enhanced/SequenceStyleGenerator.java:
```diff
@@ -234,6 +237,37 @@ public void configure(Type type, Properties params, ServiceRegistry serviceRegis
+					case FIX:
+						incrementSize = dbIncrementValue;
+					case LOG:
+						LOG.sequenceIncrementSizeMismatch( databaseSequenceName, incrementSize, dbIncrementValue );
```
dbIncrementValue is not used after this if code block and the log which says "The database sequence increment size will take precedence to avoid identifier allocation conflicts" seems incorrect

typedt commented:
> @gsmet If the allocation size is 50 in Hibernate but the DB sequence allocation size is 1, then we indeed get that weird behavior described by this issue. Hibernate will assume that it can generate the identifiers according to pooled optimizer. But for it to work properly, we need both Hibernate and the DB to agree on the same allocation size.

I've recently upgraded to 5.4.2.Final, and encountered an issue. I found the changes in this pull request broke my app. Can you please help answering my question which is, does hibernate support a use case where the DB sequence increment size is 2, and the allocationSize of the SequenceGenerator is 1? i.e, the entity id increases by 2, instead of 1, and the optimizer does not assume a pool of continuous id numbers.

vladmihalcea commented:
The DB increment size should be the same with the JPA `@SequenceGenerator` allocation size. The only time when those could differ is for the [HiLo generator](https://vladmihalcea.com/the-hilo-algorithm/), but you shouldn't use that. The default [pooled or pooled-lo](https://vladmihalcea.com/hibernate-hidden-gem-the-pooled-lo-optimizer/) are much better options and the increment size must be equal to the identifier allocation size.
